_posts/2019-04-16-preservation.md

```diff
@@ -0,0 +1,358 @@
+   browsers, and will be understood by partner institutions.
+1. This structure just needs a file system to store itself and support
+   versioning, there's no need for an object store's feature-set.
+1. We can de-duplicate as part of generating the `inventory.json` if we'd like
+   to, using up less space than the other options.
+
+**Cons**:
+
+1. The inventory file acts as a pessimistic lock in the same way as BagIt's
+   manifest file.
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the inventory will be
+   corrupted by a race condition, and versions will be unable to be properly
+   calculated:
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved as a new version of the bag, updating the
+      `inventory.json` and incrementing the version counter.
+   5. Initial preservation ends. Updates inventory, but has out-of-date checksum
+      information.
+   6. However, since versions are numbered we could theoretically check to see
+      before we start writing things if files exist to reduce the time we need
+      to lock.
+
+#### Google Cloud Storage
+
+A small note about GCS - I'm going to call paths "directories", but in GCS it's
+  all just naming semantics. `foo/bar/txt.yml` isn't a file in the directory
+  `bar`, it's a file that has a name which happens to have slashes in it. The
+  GCS tools just visualize them as directories to be nice.
+
+1. Create a base directory named `<book-id>`
+1. Upload JSON for base object to `<book-id>/<book-id>.json`
+1. If there are any binary files attached, upload them to a `data` directory,
+   like `<book-id>/data/<file-id>-<file-name>.<file-extension>`
+1. For every child, recursively start at step 1, but start in a directory named
+   after its hierarchy. IE, for a page, it would go into
+   `<book-id>/data/<page-id>'
+
+**Pros**:
+
+1. Because there is no parent "manifest" that all children have to update
+   through, we can parallelize uploads of content to preservation no matter
+   where they go over an arbitrary number of machines.
+1. Race conditions can be handled via Figgy's optimistic locking mechanism for
+   each individual resource.
+1. Versioning and checksums are a property of GCS. We just enable it on the
+   bucket and don't have to do anything special.
+
+**Cons**:
+
+1. Prioritizing parallelization makes de-duplicating infeasible.
+2. The format's not standard. It's easy to look at, but we can't point at a
+   specification. At least it should be easy to convert to any of the other
+   formats if we'd like to.
+3. Migrating to a different platform is potentially more than a single copy
+   operation - but at least GCS validates file integrity as part of the
+   platform.
```

Comment:
Are you also locked into GCS? Meaning what would migration to a new storage infrastructure look like in the future? How would you capture all of the changes you've made when you migrate to this new system since you're relying on GCS for versioning?

Reply:
I'm of the opinion we're not. Their versioning API is clearly documented and easily accessible from a variety of tools maintained both by google and third parties. If we wanted move to OCFL we could easily create a manifest, if we moved to something else that supported versions we could put them in there.

Reply:
The proposal above seems to work today. The nuances of relying on any third party storage provider for storing your audit trails is something that might need further consideration.

Reply:
I added a con below this note regarding the fact that we'd have to use their version API. I think whether or not that's a good idea can probably be less a subject of the post and more some discussion on https://github.com/pulibrary/figgy/issues/2884 - does that seem okay?